api/hasura/actions/_handlers/backfillPgive.ts

```diff
@@ -0,0 +1,418 @@
+              epoch_id: epoch.id,
+            });
+          });
+        });
+      });
+
+      if (epochObjects.length) {
+        ops[`u${circle.id}_epoch_data`] = {
+          insert_epoch_pgive_data: [
+            {
+              objects: epochObjects,
+            },
+            { __typename: true },
+          ],
+        };
+      }
+    }
+  });
+
+  if (memberObjects.length) {
+    ops[`member_data`] = {
+      insert_member_epoch_pgives: [
+        {
+          objects: memberObjects,
+        },
+        { __typename: true },
+      ],
+    };
+  }
+  if (Object.keys(ops).length) {
+    await adminClient.mutate(
+      {
+        __alias: {
+          ...ops,
+        },
+      },
+      {
+        operationName: 'insert_pGiveData',
+      }
+    );
+  }
+  res.status(200).json({ success: true });
+}
+
+type EpochData = GraphQLTypes['epoch_pgive_data_insert_input'] &
+  getCircleGiftsResult[0]['epochs'][0] & { normalizedEpochs: Array<number> };
+
+interface EpochIndexed {
+  epochs: Array<EpochData>;
+  activeMonths: number;
+  totalOptOutPgiveAlloc: Record<number, number>;
+}
+
+type getCircleGiftsResult = Awaited<ReturnType<typeof getCircleGifts>>;
+const getCircleGifts = async (limit: number) => {
+  const firstDayOfThisMonth = DateTime.local().startOf('month');
+  const { circles } = await adminClient.query({
+    circles: [
+      {
+        limit,
+        order_by: [{ id: order_by.asc }],
+        where: {
+          _and: [
+            {
+              _not: {
+                epochs: {
+                  pgive_data: {},
+                },
+              },
+            },
+            {
+              epochs: {
+                ended: { _eq: true },
+                end_date: { _lt: firstDayOfThisMonth.toISO() },
+              },
+            },
+          ],
+        },
+      },
+      {
+        id: true,
+        name: true,
+        epochs: [
+          {
+            where: {
+              ended: { _eq: true },
+              end_date: { _lt: firstDayOfThisMonth.toISO() },
+            },
+          },
+          {
+            id: true,
+            end_date: true,
+            token_gifts: [
+              {},
+              {
+                sender_id: true,
+                sender_address: true,
+                recipient_id: true,
+                recipient_address: true,
+                recipient: {
+                  name: true,
+                },
+                tokens: true,
+                note: true,
+              },
+            ],
+          },
+        ],
+        users: [
+          {},
+          {
+            id: true,
+            name: true,
+            deleted_at: true,
+            created_at: true,
+          },
+        ],
+      },
+    ],
+  });
+  return circles;
+};
+
+const roundNumbers = (value: number) => {
+  return Math.round(value * 100) / 100;
+};
+
+export default verifyHasuraRequestMiddleware(handler);
```

Comment:
Anyone can call this? Should be admin only

Reply:
I'm thinking if this takes a long time it could be a DDOS vector if anyone can call it, so maybe some scheme for restricting who can invoke this to protect against that.

Reply:
have moved it to a cron task so this should be solved